src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs

```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
+            var typeValidator = new TypeValidator(ReflectHelper);
+            var testMethodValidator = new TestMethodValidator(ReflectHelper);
+
+            return new TypeEnumerator(type, assemblyFileName, ReflectHelper, typeValidator, testMethodValidator);
+        }
+
+        private IEnumerable<UnitTestElement> DiscoverTestsInType(string assemblyFileName, string runSettingsXml, Assembly assembly, Type type, List<string> warningMessages)
+        {
+            var sourceLevelParameters = PlatformServiceProvider.Instance.SettingsProvider.GetProperties(assemblyFileName);
+            sourceLevelParameters = RunSettingsUtilities.GetTestRunParameters(runSettingsXml)?.ConcatWithOverwrites(sourceLevelParameters)
+                ?? sourceLevelParameters
+                ?? new Dictionary<string, object>();
+
+            string typeFullName = null;
+            var tests = new List<UnitTestElement>();
+
+            try
+            {
+                typeFullName = type.FullName;
+                var unitTestCases = this.GetTypeEnumerator(type, assemblyFileName).Enumerate(out var warningsFromTypeEnumerator);
+                var typeIgnored = ReflectHelper.IsAttributeDefined(type, typeof(UTF.IgnoreAttribute), false);
+
+                if (warningsFromTypeEnumerator != null)
+                {
+                    warningMessages.AddRange(warningsFromTypeEnumerator);
+                }
+
+                if (unitTestCases != null)
+                {
+                    foreach (var test in unitTestCases)
+                    {
+                        test.Ignored = typeIgnored || test.Ignored;
+                        if (test.Ignored)
+                        {
+                            tests.Add(test);
+                            continue;
+                        }
+
+                        if (this.DynamicDataAttached(sourceLevelParameters, assembly, test, tests))
+                        {
+                            continue;
+                        }
+
+                        tests.Add(test);
+                    }
+                }
+            }
+            catch (Exception exception)
+            {
+                // If we fail to discover type from a class, then don't abort the discovery
+                // Move to the next type.
+                string message = string.Format(CultureInfo.CurrentCulture, Resource.CouldNotInspectTypeDuringDiscovery, typeFullName, assemblyFileName, exception.Message);
+                PlatformServiceProvider.Instance.AdapterTraceLogger.LogInfo($"AssemblyEnumerator: {message}");
+                warningMessages.Add(message);
+            }
+
+            return tests;
+        }
+
+        private bool DynamicDataAttached(IDictionary<string, object> sourceLevelParameters, Assembly assembly, UnitTestElement test, List<UnitTestElement> tests)
+        {
+            if (test.TestMethod.HasManagedMethodAndTypeProperties == false)
```

Comment:
It should always be `true`, but any part of the chain is obsolete; it might not contain those. Since we depend on those properties, if they don't exist, we bail out early—added comment.